File: Sources/HoverPocket/Capabilities/CalculatorCapability.swift
import Foundation

struct CalculatorCapabilityEvaluation: Equatable, Sendable {
    let normalizedExpression: String
    let result: String
}

enum CalculatorCapabilityEvaluator {
    private enum Operation: Character {
        case add = "+"
        case subtract = "-"
        case multiply = "*"
        case divide = "/"
    }

    private static let maximumMagnitude = Decimal(string: "999999999999999999", locale: posixLocale)!
    private static let posixLocale = Locale(identifier: "en_US_POSIX")
    private static let maximumValues = 64

    static func evaluate(_ expression: String) throws -> CalculatorCapabilityEvaluation {
        let normalized = expression
            .replacingOccurrences(of: "−", with: "-")
            .replacingOccurrences(of: "×", with: "*")
            .replacingOccurrences(of: "÷", with: "/")
            .replacingOccurrences(of: ",", with: ".")
            .unicodeScalars
            .filter { !CharacterSet.whitespacesAndNewlines.contains($0) }
            .map(String.init)
            .joined()
        guard !normalized.isEmpty,
              normalized.unicodeScalars.count <= 256,
              normalized.unicodeScalars.allSatisfy({ $0.isASCII }) else {
            throw CapabilityHandlerError.invalidArgument("expression")
        }

        let characters = Array(normalized)
        var values: [Decimal] = []
        var operations: [Operation] = []
        var index = 0
        while index < characters.count {
            guard values.count < maximumValues else {
                throw CapabilityHandlerError.invalidArgument("expression")
            }
            let start = index
            if characters[index] == "+" || characters[index] == "-" {
                index += 1
            }
            var integerDigits = 0
            var fractionDigits = 0
            var sawDecimal = false
            while index < characters.count {
                let character = characters[index]
                if character.isNumber {
                    if sawDecimal {
                        fractionDigits += 1
                    } else {
                        integerDigits += 1
                    }
                    index += 1
                    continue
                }
                if character == ".", !sawDecimal {
                    sawDecimal = true
                    index += 1
                    continue
                }
                break
            }
            guard integerDigits + fractionDigits > 0,
                  integerDigits <= 18,
                  fractionDigits <= 12,
                  index - start <= 32,
                  let value = Decimal(
                    string: String(characters[start..<index]),
                    locale: posixLocale
                  ),
                  !value.isNaN,
                  value.magnitude <= maximumMagnitude else {
                throw CapabilityHandlerError.invalidArgument("expression")
            }
            values.append(value)
            if index == characters.count {
                break
            }
            guard let operation = Operation(rawValue: characters[index]) else {
                throw CapabilityHandlerError.invalidArgument("expression")
            }
            operations.append(operation)
            index += 1
            guard index < characters.count else {
                throw CapabilityHandlerError.invalidArgument("expression")
            }
        }
        guard values.count == operations.count + 1 else {
            throw CapabilityHandlerError.invalidArgument("expression")
        }

        var reducedValues = [values[0]]
        var reducedOperations: [Operation] = []
        for (operationIndex, operation) in operations.enumerated() {
            let rhs = values[operationIndex + 1]
            switch operation {
            case .multiply, .divide:
                guard let lhs = reducedValues.popLast() else {
                    throw CapabilityHandlerError.invalidArgument("expression")
                }
                reducedValues.append(try calculate(lhs, rhs, operation))
            case .add, .subtract:
                reducedOperations.append(operation)
                reducedValues.append(rhs)
            }
        }
        var result = reducedValues[0]
        for (operationIndex, operation) in reducedOperations.enumerated() {
            result = try calculate(result, reducedValues[operationIndex + 1], operation)
        }
        var normalizedParts: [String] = []
        for valueIndex in values.indices {
            normalizedParts.append(format(values[valueIndex]))
            if valueIndex < operations.count {
                normalizedParts.append(String(operations[valueIndex].rawValue))
            }
        }
        return CalculatorCapabilityEvaluation(
            normalizedExpression: normalizedParts.joined(separator: " "),
            result: format(result)
        )
    }

    private static func calculate(
        _ lhs: Decimal,
        _ rhs: Decimal,
        _ operation: Operation
    ) throws -> Decimal {
        if operation == .divide, rhs == 0 {
            throw CapabilityHandlerError.invalidArgument("expression")
        }
        let raw: Decimal
        switch operation {
        case .add: raw = lhs + rhs
        case .subtract: raw = lhs - rhs
        case .multiply: raw = lhs * rhs
        case .divide: raw = lhs / rhs
        }
        guard !raw.isNaN, raw.magnitude <= maximumMagnitude else {
            throw CapabilityHandlerError.invalidArgument("expression")
        }
        var input = raw
        var rounded = Decimal()
        NSDecimalRound(&rounded, &input, 12, .plain)
        guard !rounded.isNaN, rounded.magnitude <= maximumMagnitude else {
            throw CapabilityHandlerError.invalidArgument("expression")
        }
        return rounded
    }

    private static func format(_ value: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.locale = posixLocale
        formatter.numberStyle = .decimal
        formatter.usesGroupingSeparator = false
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = 12
        formatter.roundingMode = .halfUp
        return formatter.string(from: NSDecimalNumber(decimal: value)) ?? "0"
    }
}

@MainActor
final class CalculatorEvaluateCapabilityHandler: PocketCapabilityHandler {
    let key = PocketCapabilityKeys.calculatorEvaluate

    func handle(
        arguments: CapabilityObject,
        context: CapabilityHandlerContext
    ) async throws -> CapabilityObject {
        _ = context
        guard Set(arguments.keys) == ["expression"] else {
            throw CapabilityHandlerError.invalidArgument("expression")
        }
        let evaluation = try CalculatorCapabilityEvaluator.evaluate(
            arguments.requiredString("expression", maxLength: 256)
        )
        return [
            "normalizedExpression": .string(evaluation.normalizedExpression),
            "result": .string(evaluation.result)
        ]
    }
}

File: windows/src/HoverPocket.Shell/Capabilities/CalculatorCapability.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace HoverPocket.Shell.Capabilities;

internal sealed record CalculatorCapabilityEvaluation(string NormalizedExpression, string Result);

internal static class CalculatorCapabilityEvaluator
{
    private const int MaximumValues = 64;
    private const decimal MaximumMagnitude = 999999999999999999m;

    public static CalculatorCapabilityEvaluation Evaluate(string expression)
    {
        var normalized = string.Concat(expression
            .Replace('−', '-')
            .Replace('×', '*')
            .Replace('÷', '/')
            .Replace(',', '.')
            .Where(character => !char.IsWhiteSpace(character)));
        if (normalized.Length is 0 or > 256 || normalized.Any(character => character > 0x7f))
        {
            throw CapabilityJson.Invalid("expression");
        }

        var values = new List<decimal>();
        var operations = new List<char>();
        var index = 0;
        while (index < normalized.Length)
        {
            if (values.Count >= MaximumValues)
            {
                throw CapabilityJson.Invalid("expression");
            }
            var start = index;
            if (normalized[index] is '+' or '-')
            {
                index += 1;
            }
            var integerDigits = 0;
            var fractionDigits = 0;
            var sawDecimal = false;
            while (index < normalized.Length)
            {
                var character = normalized[index];
                if (character is >= '0' and <= '9')
                {
                    if (sawDecimal)
                    {
                        fractionDigits += 1;
                    }
                    else
                    {
                        integerDigits += 1;
                    }
                    index += 1;
                    continue;
                }
                if (character == '.' && !sawDecimal)
                {
                    sawDecimal = true;
                    index += 1;
                    continue;
                }
                break;
            }
            if (integerDigits + fractionDigits == 0
                || integerDigits > 18
                || fractionDigits > 12
                || index - start > 32
                || !decimal.TryParse(
                    normalized.AsSpan(start, index - start),
                    NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint,
                    CultureInfo.InvariantCulture,
                    out var value)
                || Math.Abs(value) > MaximumMagnitude)
            {
                throw CapabilityJson.Invalid("expression");
            }
            values.Add(value);
            if (index == normalized.Length)
            {
                break;
            }
            var operation = normalized[index];
            if (operation is not ('+' or '-' or '*' or '/'))
            {
                throw CapabilityJson.Invalid("expression");
            }
            operations.Add(operation);
            index += 1;
            if (index == normalized.Length)
            {
                throw CapabilityJson.Invalid("expression");
            }
        }
        if (values.Count != operations.Count + 1)
        {
            throw CapabilityJson.Invalid("expression");
        }

        var reducedValues = new List<decimal> { values[0] };
        var reducedOperations = new List<char>();
        for (var operationIndex = 0; operationIndex < operations.Count; operationIndex++)
        {
            var operation = operations[operationIndex];
            var rhs = values[operationIndex + 1];
            if (operation is '*' or '/')
            {
                var lhs = reducedValues[^1];
                reducedValues[^1] = Calculate(lhs, rhs, operation);
            }
            else
            {
                reducedOperations.Add(operation);
                reducedValues.Add(rhs);
            }
        }
        var result = reducedValues[0];
        for (var operationIndex = 0; operationIndex < reducedOperations.Count; operationIndex++)
        {
            result = Calculate(result, reducedValues[operationIndex + 1], reducedOperations[operationIndex]);
        }

        var builder = new StringBuilder();
        for (var valueIndex = 0; valueIndex < values.Count; valueIndex++)
        {
            if (valueIndex > 0)
            {
                builder.Append(' ').Append(operations[valueIndex - 1]).Append(' ');
            }
            builder.Append(Format(values[valueIndex]));
        }
        return new CalculatorCapabilityEvaluation(builder.ToString(), Format(result));
    }

    private static decimal Calculate(decimal lhs, decimal rhs, char operation)
    {
        if (operation == '/' && rhs == 0m)
        {
            throw CapabilityJson.Invalid("expression");
        }
        try
        {
            var raw = operation switch
            {
                '+' => checked(lhs + rhs),
                '-' => checked(lhs - rhs),
                '*' => checked(lhs * rhs),
                '/' => checked(lhs / rhs),
                _ => throw CapabilityJson.Invalid("expression")
            };
            if (Math.Abs(raw) > MaximumMagnitude)
            {
                throw CapabilityJson.Invalid("expression");
            }
            return Math.Round(raw, 12, MidpointRounding.AwayFromZero);
        }
        catch (OverflowException)
        {
            throw CapabilityJson.Invalid("expression");
        }
    }

    private static string Format(decimal value) =>
        value.ToString("0.############", CultureInfo.InvariantCulture);
}

internal sealed class CalculatorEvaluateCapabilityHandler : IPocketCapabilityHandler
{
    public PocketCapabilityKey Key => CapabilityIds.CalculatorEvaluate;

    public Task<JsonElement> HandleAsync(
        JsonElement arguments,
        CapabilityHandlerContext context,
        CancellationToken cancellationToken = default)
    {
        _ = context;
        cancellationToken.ThrowIfCancellationRequested();
        CapabilitySchemaValidation.ExactKeys(arguments, ["expression"]);
        var evaluation = CalculatorCapabilityEvaluator.Evaluate(
            CapabilityJson.RequiredString(arguments, "expression", 256));
        return Task.FromResult(CapabilityJson.From(new
        {
            evaluation.NormalizedExpression,
            evaluation.Result
        }));
    }
}
